src/c++/perf_analyzer/perf_utils.cc

```diff
@@ -423,4 +425,292 @@ ScheduleDistribution<Distribution::CONSTANT>(const double request_rate)
+void
+MPI_BcastIntWorld(void* buffer, int count, int root)
+{
+  void* handle{GetMPIHandle()};
+  if (handle == nullptr) {
+    throw std::runtime_error(
+        "Unable to load MPI library. Check that 'libmpi.so' is on "
+        "`LD_LIBRARY_PATH` environment variable path.");
+  }
+
+  MPIImplementation implementation{GetMPIImplementation()};
+  if (implementation == MPIImplementation::MPICH) {
+    int (*MPI_Bcast)(void*, int, int, int, int){
+        (int (*)(void*, int, int, int, int))dlsym(handle, "MPI_Bcast")};
+    if (MPI_Bcast == nullptr) {
+      throw std::runtime_error(
+          "Unable to obtain address of `MPI_Bcast` symbol.");
+    }
+
+    MPI_Bcast(buffer, count, MPI_INTMPICH(), root, MPI_COMM_WORLDMPICH());
+  } else if (implementation == MPIImplementation::OpenMPI) {
+    int (*MPI_Bcast)(void*, int, void*, int, void*){
+        (int (*)(void*, int, void*, int, void*))dlsym(handle, "MPI_Bcast")};
+    if (MPI_Bcast == nullptr) {
+      throw std::runtime_error(
+          "Unable to obtain address of `MPI_Bcast` symbol.");
+    }
+
+    MPI_Bcast(buffer, count, MPI_INTOpenMPI(), root, MPI_COMM_WORLDOpenMPI());
+  }
+}
+
+void
+MPI_Finalize()
+{
+  void* handle{GetMPIHandle()};
+  if (handle == nullptr) {
+    return;
+  }
+
+  int (*MPI_Finalize)(){(int (*)())dlsym(handle, "MPI_Finalize")};
+  if (MPI_Finalize == nullptr) {
+    throw std::runtime_error(
+        "Unable to obtain address of `MPI_Finalize` symbol.");
+  }
+
+  MPI_Finalize();
+}
+
+void*
+MPI_COMM_WORLDOpenMPI()
+{
+  void* handle{GetMPIHandle()};
+  if (handle == nullptr) {
+    throw std::runtime_error(
+        "Unable to load MPI library. Check that 'libmpi.so' is on "
+        "`LD_LIBRARY_PATH` environment variable path.");
+  }
+  void* MPI_COMM_WORLD{dlsym(handle, "ompi_mpi_comm_world")};
+  if (MPI_COMM_WORLD == nullptr) {
+    throw std::runtime_error(
+        "Unable to obtain address of `ompi_mpi_comm_world` symbol.");
+  }
+  return MPI_COMM_WORLD;
+}
+
+int
+MPI_COMM_WORLDMPICH()
+{
+  int MPI_COMM_WORLD{0x44000000};
+  return MPI_COMM_WORLD;
+}
+
+void*
+MPI_INTOpenMPI()
+{
+  void* handle{GetMPIHandle()};
+  if (handle == nullptr) {
+    throw std::runtime_error(
+        "Unable to load MPI library. Check that 'libmpi.so' is on "
+        "`LD_LIBRARY_PATH` environment variable path.");
+  }
+  void* MPI_INT{dlsym(handle, "ompi_mpi_int")};
+  if (MPI_INT == nullptr) {
+    throw std::runtime_error(
+        "Unable to obtain address of `ompi_mpi_int` symbol.");
+  }
+  return MPI_INT;
+}
+
+int
+MPI_INTMPICH()
+{
+  int MPI_INT{0x4c000405};
+  return MPI_INT;
+}
+
+void*
+GetMPIHandle()
+{
+  return dlopen("libmpi.so", RTLD_LAZY | RTLD_GLOBAL);
+}
+
+bool
+IsMPIRun()
+{
+  if (GetMPIHandle() == nullptr) {
+    return false;
+  }
+
+  if (MPI_Initialized() == false) {
+    throw std::runtime_error("Must call MPI_Init() before calling IsMPIRun().");
+  }
+
+  return MPI_Comm_sizeWorld() > 1;
+}
+
+MPIImplementation
+GetMPIImplementation()
```

Comment:
Can we encapsulate all these MPI related functions in a class MPIDriver? We would just need to do GetMPIHandle()(read dlopen) once then keep on reusing the handle for all these future calls.

Reply:
Encapsulated in `MPIDriver` class 👍